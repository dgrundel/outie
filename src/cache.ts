
export interface Cache<T> {
    has: (key: string) => boolean;
    put: (key: string, value: T) => void;
    get: (key: string) => Promise<T | undefined>;
    getOrPut: (key: string, supplier: () => Promise<T>) => Promise<T>;
}

export class MruCache<T> implements Cache<T> {
    private readonly maxSize: number;
    private readonly items: Record<string, T>;
    private readonly keys: Set<string>;

    constructor(maxSize: number = 150) {
        this.maxSize = maxSize;
        this.items = {};
        this.keys = new Set();
    }

    has(key: string): boolean {
        return this.keys.has(key);
    }

    put(key: string, value: T): void {
        if (this.has(key)) {
            throw new Error(`Duplicate cache key "${key}"`);
        }

        this.resize();

        this.keys.add(key);
        this.items[key] = value;
    }

    async get(key: string): Promise<T | undefined> {
        if (this.keys.has(key)) {
            this.touchKey(key);
            return this.items[key];
        }

        return undefined;
    }

    async getOrPut(key: string, supplier: () => Promise<T>): Promise<T> {
        if (this.has(key)) {
            return (await this.get(key))!;
        }

        const value: T = await supplier();
        this.put(key, value);
        return value;
    }

    /**
     * Move a cache key to the end of the
     * deletion queue.
     * @param key cache key
     */
    private touchKey(key: string): void {
        if (!this.keys.has(key)) {
            throw new Error(`Key "${key}" does not exist.`);
        }

        this.keys.delete(key);
        this.keys.add(key);
    }

    /**
     * Remove items to make room for new ones.
     * Deletes oldest keys first.
     */
    private resize(): void {
        // delete keys in order of insertion
        while (this.keys.size >= this.maxSize) {
            const iter = this.keys[Symbol.iterator]();
            const key = iter.next().value;
            
            this.keys.delete(key);
            delete this.items[key];
        }
    }
}