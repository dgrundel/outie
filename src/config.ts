export interface OutieConfig {
    tokenStart: string;
    tokenEnd: string;
    rawTokenIdentifier: string;
    includeTokenIdentifier: string;
}

const defaultConfig: OutieConfig = {
    tokenStart: '{',
    tokenEnd: '}',
    rawTokenIdentifier: 'raw',
    includeTokenIdentifier: 'include',
};

export const getConfig = (userConfig: Partial<OutieConfig> = {}) => {
    return Object.assign({}, defaultConfig, userConfig);
};

export type RenderModel = Record<string, any>;