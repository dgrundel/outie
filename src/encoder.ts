// https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html

export const encodeHtml = (s: string): string => {
    return s
        // & --> &amp;
        .replace(/\&/g, '&amp;')
        // < --> &lt;
        .replace(/\</g, '&lt;')
        // > --> &gt;
        .replace(/\>/g, '&gt;')
        // " --> &quot;
        .replace(/\"/g, '&#x22;')
        // ' --> &#x27;
        .replace(/\'/g, '&#x27;');
};