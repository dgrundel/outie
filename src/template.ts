import * as fs from 'fs';
import * as path from 'path';
import { RenderModel } from './config';

export class Template {
    readonly content: string;
    readonly model: RenderModel;
    readonly dir?: string;

    constructor(content: string, model: RenderModel, dir?: string) {
        this.content = content;
        this.model = model;
        this.dir = dir;
    }

    static async fromFile(filePath: string, model: RenderModel, cwd?: string): Promise<Template> {
        const absPath = path.isAbsolute(filePath) || typeof cwd !== 'string'
            ? filePath
            : path.join(cwd, filePath);
        const contents: string = await new Promise((resolve, reject) => {
            fs.readFile(path.resolve(absPath), 'utf8', (err, data) => err ? reject(err) : resolve(data));
        });

        return new Template(contents, model, path.dirname(absPath));
    }

    static async fromString(content: string, model: RenderModel): Promise<Template> {
        return new Template(content, model);
    }
}