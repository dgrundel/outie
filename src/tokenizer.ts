import { OutieConfig } from "./config";
import { BlockEndToken } from './tokens/core/BlockEndToken';
import { BlockStartToken } from './tokens/core/BlockStartToken';
import { ModelKeyToken } from './tokens/ModelKeyToken';
import { RawToken } from './tokens/RawToken';
import { Token, TokenConstructor } from './tokens/core/Token';
import { RootToken } from './tokens/RootToken';
import { Template } from './template';

const identInfo = (s: string, identifier: string) => {
    const value = s.trim();
    const present = value.startsWith(identifier);
    const trimmed = present
        ? value.substring(identifier.length).trim()
        : value;
    return { present, trimmed };
};

export class Tokenizer {
    readonly config: OutieConfig;

    constructor(config: OutieConfig) {
        this.config = config;
    }

    createToken (content: string, sourceTemplate?: Template): Token {
        const config = this.config;
        // this lops off the opening "/" sequence if present
        const { present: isClosingToken, trimmed } = identInfo(content, config.closeTokenIdentifier);

        // sort identifiers by length, longest first
        // this avoids cases where we might interpret "includeRaw" as "include"
        // because it starts with the same string ('include')
        const orderedIdentifiers = Object.keys(config.tokens).sort((a, b) => {
            if (a.length === b.length) {
                return 0;
            }
            return a.length > b.length ? -1 : 1;
        });

        // test for each identifier
        for (let i = 0; i < orderedIdentifiers.length; i++) {
            const identifier = orderedIdentifiers[i];
            const { present: isMatchingType, trimmed: tokenContents } = identInfo(trimmed, identifier);
            
            // found a match, create a token
            if (isMatchingType) {
                const TokenType: TokenConstructor = config.tokens[identifier];
                return isClosingToken
                    ? new BlockEndToken(TokenType)
                    : new TokenType(tokenContents, sourceTemplate);
            }
        }

        return new ModelKeyToken(content);
    };

    tokenize (s: string, sourceTemplate?: Template): Token[] {
        const root = new RootToken();
        const stack: BlockStartToken[] = [];
        let i = 0;
    
        while (i < s.length) {
            const start = s.indexOf(this.config.tokenStart, i);
            if (start === -1) {
                break;
            }
            const end = s.indexOf(this.config.tokenEnd, start + this.config.tokenStart.length);
            if (end === -1) {
                break;
            }
            
            const parent: BlockStartToken = stack.length ? stack[stack.length - 1] : root;

            if (i < start) {
                parent.append(new RawToken(s.substring(i, start)));
            }
            
            const token = this.createToken(
                s.substring(start + this.config.tokenStart.length, end),
                sourceTemplate
            );

            if (token instanceof BlockEndToken) {
                if (stack.length === 0) {
                    throw new Error(`Found closing ${token.type.name} without opening.`);
                }
                const openingToken = stack[stack.length - 1];
                if (openingToken.constructor !== token.type) {
                    throw new Error(`Found ${token.type.name} but ${openingToken.constructor.name} is still open.`);
                }

                stack.pop();

            } else if (token instanceof BlockStartToken) {
                parent.append(token);
                stack.push(token);

            } else {
                parent.append(token);
            }
    
            i = end + this.config.tokenEnd.length;
        }

        if (stack.length > 0) {
            throw new Error(`Unclosed items in template: ${stack.map(t => t.constructor.name).join(', ')}`);
        }

        if (i < s.length) {
            root.append(new RawToken(s.substring(i)));
        }
    
        return [root];
    }
}