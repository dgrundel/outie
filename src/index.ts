import { getConfig, OutieConfig, RenderModel } from "./config";
import { Tokenizer } from "./tokenizer";
import { Template } from "./template";

export class Outie {
    readonly config: OutieConfig;
    readonly tokenizer: Tokenizer;

    constructor(userConfig?: Partial<OutieConfig>) {
        const config = getConfig(userConfig);
        
        this.config = config;
        this.tokenizer = new Tokenizer(config);
    }

    async render (template: string, model: RenderModel): Promise<string> {
        const t = await Template.fromString(template, model);
        return this.tokenizer.renderTemplate(t);
    }

    async renderFile (filePath: string, model: RenderModel): Promise<string> {
        const t = await Template.fromFile(filePath, model);
        return this.tokenizer.renderTemplate(t);
    }
}