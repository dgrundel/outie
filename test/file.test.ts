import { Outie } from "../src";
import * as path from 'path';

describe('outie file handling', () => {
    it('should render a basic template from a file', async () => {
        const outie = new Outie();
        const filePath = path.resolve(__dirname, '__testData__/simple.html.outie');
        const actual = await outie.renderFile(filePath, { name: 'world' });
        
        expect(actual).toMatchSnapshot();
    });

    it('should handle relative includes', async () => {
        const outie = new Outie();
        const filePath = path.resolve(__dirname, '__testData__/include.html.outie');
        const actual = await outie.renderFile(filePath, { name: 'Kate' });
        
        expect(actual).toMatchSnapshot();
    });

    it('raw includes are unprocessed', async () => {
        const outie = new Outie();
        const filePath = path.resolve(__dirname, '__testData__/rawInclude.html.outie');
        const actual = await outie.renderFile(filePath, { name: 'Kate' });
        
        expect(actual).toMatchSnapshot();
    });
});