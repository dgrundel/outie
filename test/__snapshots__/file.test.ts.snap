// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outie file handling should handle relative includes 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <meta charset=\\"UTF-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <title>Test Page</title>
</head>
<body>
    <h1>Hello, Kate!</h1>
    <h2>Here's a partial for you, Kate!</h2>
</body>
</html>"
`;

exports[`outie file handling should render a basic template from a file 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <meta charset=\\"UTF-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <title>Test Page</title>
</head>
<body>
    <h1>Hello, world!</h1>
</body>
</html>"
`;
